pkg/cvo/configuration/configuration_test.go

```diff
@@ -0,0 +1,276 @@
+				ObjectMeta: metav1.ObjectMeta{
+					Generation: 4,
+				},
+				Spec: operatorv1alpha1.ClusterVersionOperatorSpec{
+					OperatorLogLevel: operatorv1.Trace,
+				},
+				Status: operatorv1alpha1.ClusterVersionOperatorStatus{
+					ObservedGeneration: 4,
+				},
+			},
+			internalConfig: ClusterVersionOperatorConfiguration{
+				desiredLogLevel:        operatorv1.Normal,
+				lastObservedGeneration: 3,
+			},
+			expectedInternalConfig: ClusterVersionOperatorConfiguration{
+				desiredLogLevel:        operatorv1.Trace,
+				lastObservedGeneration: 4,
+			},
+		},
+		{
+			name: "number of not observed generations does not impact sync",
+			config: operatorv1alpha1.ClusterVersionOperator{
+				ObjectMeta: metav1.ObjectMeta{
+					Generation: 40,
+				},
+				Spec: operatorv1alpha1.ClusterVersionOperatorSpec{
+					OperatorLogLevel: operatorv1.TraceAll,
+				},
+				Status: operatorv1alpha1.ClusterVersionOperatorStatus{
+					ObservedGeneration: 3,
+				},
+			},
+			expectedConfig: operatorv1alpha1.ClusterVersionOperator{
+				ObjectMeta: metav1.ObjectMeta{
+					Generation: 40,
+				},
+				Spec: operatorv1alpha1.ClusterVersionOperatorSpec{
+					OperatorLogLevel: operatorv1.TraceAll,
+				},
+				Status: operatorv1alpha1.ClusterVersionOperatorStatus{
+					ObservedGeneration: 40,
+				},
+			},
+			internalConfig: ClusterVersionOperatorConfiguration{
+				desiredLogLevel:        operatorv1.Normal,
+				lastObservedGeneration: 3,
+			},
+			expectedInternalConfig: ClusterVersionOperatorConfiguration{
+				desiredLogLevel:        operatorv1.TraceAll,
+				lastObservedGeneration: 40,
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			// Initialize testing logic
+			client := operatorclientsetfake.NewClientset(&tt.config)
+			tt.internalConfig.client = client.OperatorV1alpha1().ClusterVersionOperators()
+			ctx, cancelFunc := context.WithDeadline(context.Background(), time.Now().Add(time.Minute))
+
+			// Run tested functionality
+			if err := tt.internalConfig.sync(ctx, &tt.config); err != nil {
+				t.Errorf("unexpected error %v", err)
+			}
+
+			// Verify results
+			if tt.internalConfig.lastObservedGeneration != tt.expectedInternalConfig.lastObservedGeneration {
+				t.Errorf("unexpected 'lastObservedGeneration' value; wanted=%v, got=%v", tt.expectedInternalConfig.lastObservedGeneration, tt.internalConfig.lastObservedGeneration)
+			}
+			if tt.internalConfig.desiredLogLevel != tt.expectedInternalConfig.desiredLogLevel {
+				t.Errorf("unexpected 'desiredLogLevel' value; wanted=%v, got=%v", tt.expectedInternalConfig.desiredLogLevel, tt.internalConfig.desiredLogLevel)
+			}
+
+			config, err := client.OperatorV1alpha1().ClusterVersionOperators().Get(ctx, "", metav1.GetOptions{})
+			if err != nil {
+				t.Errorf("unexpected error %v", err)
+			}
+			if diff := cmp.Diff(tt.expectedConfig, *config, cmpopts.IgnoreFields(operatorv1alpha1.ClusterVersionOperator{}, "ObjectMeta")); diff != "" {
```

Comment:
nit: just ignore the necessary?

```suggestion
			if diff := cmp.Diff(tt.expectedConfig, *config, cmpopts.IgnoreFields(metav1.ObjectMeta{}, "ManagedFields")); diff != "" {
```

Reply:
Thanks for raising this issue!

It resulted in finding that the controller does not set the `ManagedFieldsEntry[i].Manager` field.

New commit https://github.com/openshift/cluster-version-operator/pull/1170/commits/fec679998f6639ea5590bfac726afec9294926b6 sets the manager field.

The https://github.com/openshift/cluster-version-operator/pull/1170/commits/2217234aaad92ab57514b18c289ae0aa09b3e523 updates the tests.

Reply:
We simply declare the name of the manager that manages the field. By setting `fieldManager` in the `UpdateOptions` setting for the `UpdateStatus` call, which [`is a name associated with the actor or entity that is making these changes`](https://github.com/openshift/cluster-version-operator/blob/25f02d1c2b32020b20ef9a6e1e6c0b3cd4f8496a/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L721-L740). That's the benefit.

> The Kubernetes API server tracks managed fields for all newly created objects.
> 
> When trying to apply an object, fields that have a different value and are owned by another [manager](https://kubernetes.io/docs/reference/using-api/server-side-apply/#managers) will result in a [conflict](https://kubernetes.io/docs/reference/using-api/server-side-apply/#conflicts). This is done in order to signal that the operation might undo another collaborator's changes. Writes to objects with managed fields can be forced, in which case the value of any conflicted field will be overridden, and the ownership will be transferred.
> 
> Whenever a field's value does change, ownership moves from its current manager to the manager making the change.

Field management is used in Server-Side Apply: https://kubernetes.io/docs/reference/using-api/server-side-apply/

Reply:
I got the meaning of the API. However, I still do not feel it is worth doing.
To me, knowing the actor that made the change on the object is not needed.

BTW, there are no other places in the CVO repo setting that up.
I can not see anything special for this situation.
As I said before, I am not against it either if you want it (as I do not see harm either 😄 ).

```console
$ rg UpdateOptions -g '!vendor/'
pkg/updatestatus/updatestatuscontroller.go
395:    _, err = c.configMaps.Update(ctx, c.statusApi.cm, metav1.UpdateOptions{})

pkg/cvo/status.go
651:    actual, err := client.ClusterVersions().UpdateStatus(ctx, required, metav1.UpdateOptions{})
664:            actual, err = client.ClusterVersions().UpdateStatus(ctx, required, metav1.UpdateOptions{})

pkg/cvo/internal/generic.go
112:    actual, err := client.Update(ctx, existing, metav1.UpdateOptions{})

pkg/cvo/internal/operatorstatus.go
118:            if _, err := b.createClient.UpdateStatus(ctx, clusterOperator, metav1.UpdateOptions{}); err != nil {
141:                    if _, err := b.createClient.Update(ctx, existing, metav1.UpdateOptions{}); err != nil {

pkg/cvo/cvo_test.go
3849:                                           _, err = f.CoreV1().ConfigMaps("test").Update(ctx, tt.gateCm, metav1.UpdateOptions{})
3863:                                           _, err = f.CoreV1().ConfigMaps("test").Update(ctx, tt.ackCm, metav1.UpdateOptions{})

lib/resourceapply/operators.go
49:     actual, err := client.OperatorGroups(required.Namespace).Update(ctx, existing, metav1.UpdateOptions{})

lib/resourceapply/admissionregistration.go
46:     actual, err := client.ValidatingWebhookConfigurations().Update(ctx, existing, metav1.UpdateOptions{})

lib/resourceapply/imagestream.go
45:     actual, err := client.ImageStreams(required.Namespace).Update(ctx, existing, metav1.UpdateOptions{})

lib/resourceapply/cv.go
41:     actual, err := client.ClusterVersions().Update(ctx, existing, metav1.UpdateOptions{})

lib/resourceapply/batch.go
47:     actual, err := client.Jobs(required.Namespace).Update(ctx, existing, metav1.UpdateOptions{})
82:     actual, err := client.CronJobs(required.Namespace).Update(ctx, existing, metav1.UpdateOptions{})

lib/resourceapply/apiext.go
47:     actual, err := client.CustomResourceDefinitions().Update(ctx, existing, metav1.UpdateOptions{})

lib/resourceapply/security.go
44:     actual, err := client.SecurityContextConstraints().Update(ctx, reconcile, metav1.UpdateOptions{})

lib/resourceapply/apps.go
47:     actual, err := client.Deployments(required.Namespace).Update(ctx, existing, metav1.UpdateOptions{})
77:     actual, err := client.DaemonSets(required.Namespace).Update(ctx, existing, metav1.UpdateOptions{})

lib/resourceapply/rbac.go
41:     actual, err := client.ClusterRoleBindings().Update(ctx, existing, metav1.UpdateOptions{})
70:     actual, err := client.ClusterRoles().Update(ctx, existing, metav1.UpdateOptions{})
99:     actual, err := client.RoleBindings(required.Namespace).Update(ctx, existing, metav1.UpdateOptions{})
128:    actual, err := client.Roles(required.Namespace).Update(ctx, existing, metav1.UpdateOptions{})

lib/resourceapply/core.go
44:     actual, err := client.Namespaces().Update(ctx, existing, metav1.UpdateOptions{})
81:     actual, err := client.Services(required.Namespace).Update(ctx, existing, metav1.UpdateOptions{})
111:    actual, err := client.ServiceAccounts(required.Namespace).Update(ctx, existing, metav1.UpdateOptions{})
141:    actual, err := client.ConfigMaps(required.Namespace).Update(ctx, existing, metav1.UpdateOptions{})

```

Reply:
So the CVO does set the manager field for its updates. For example, https://issues.redhat.com/browse/OCPBUGS-19391.

And by setting the `fieldManager` in this PR we can override the value (I have tested this right now). I am trying to figure out where is it being set initially.

Reply:
> I'll look into the origin of the manager field value later.

I bet it is from the work of some k8s webhook.

Reply:
https://github.com/openshift/cluster-version-operator/pull/1170#discussion_r2036080371 done in https://github.com/openshift/cluster-version-operator/pull/1170/commits/0f5d0d70b1d40e576f0ee82285e94ead9a746660

Reply:
Marking as resolved. The field is dropped from the PR.